src/qibo/parallel.py

```diff
@@ -0,0 +1,193 @@
+    _instance = None
+    # dict with of shared objects
+    _objects_per_process = {}
+    custom_function = None
+    lock = None
+    arguments = ()
+
+    def __new__(cls, *args, **kwargs):
+        """Creates singleton instance."""
+        if cls._instance is None:
+            cls._instance = super(ParallelResources, cls).__new__(
+                cls, *args, **kwargs)
+        return cls._instance
+
+    def run(self, params=None):
+        """Evaluates the custom_function object for a specific set of
+        parameters. This class performs the lock mechanism to duplicate objects
+        for each process.
+        """
+        # lock to avoid race conditions
+        self.lock.acquire()
+        # get process name
+        pname = self.mp.current_process().name
+        # check if there are objects already stored
+        args = self._objects_per_process.get(pname, None)
+        if args is None:
+            args = []
+            for obj in self.arguments:
+                try:
+                    # copy object if copy method is available
+                    copy = obj.copy(deep=True)
+                except AttributeError:
+                    # if copy is not implemented just use the original object
+                    copy = obj
+                except Exception as e:
+                    # use print otherwise the message will not appear
+                    print('Exception in ParallelResources', str(e))
+                args.append(copy)
+            args = tuple(args)
+            self._objects_per_process[pname] = args
+        # unlock
+        self.lock.release()
+        # finally compute the custom function
+        return self.custom_function(params, *args)
+
+    def reset(self):
+        """Cleanup memory."""
+        self._objects_per_process = {}
+        self.custom_function = None
+        self.lock = None
+        self.arguments = ()
+
+
+def _executor(params): # pragma: no cover
+    """Executes singleton call."""
+    return ParallelResources().run(params)
+
+
+def parallel_execution(circuit, states, processes=None):
+    """Execute circuit for multiple states.
+
+    Example:
+        ::
+
+            from qibo import models, set_threads
+            from qibo.parallel import parallel_execution
+            import numpy as np
+            # create circuit
+            nqubits = 22
+            circuit = models.QFT(nqubits)
+            # create random states
+            states = [ np.random.random(2**nqubits) for i in range(5)]
+            # set threads to 1 per process (optional, requires tuning)
+            set_threads(1)
+            # execute in parallel
+            results = parallel_execution(circuit, states, processes=2)
+
+    Args:
+        circuit (qibo.models.Circuit): the input circuit.
+        states (list): list of states for the circuit evaluation.
+        processes (int): number of processes for parallel evaluation.
+
+    Returns:
+        Circuit evaluation for input states.
+    """
```

Comment:
Another useful application of this method is that it can be used to parallelize noise simulation with repetition if `None` is passed as an initial state for all circuits. For example:
```Python
nreps = 10

c = models.QFT(20)
c = c.with_noise((0.2, 0.0, 0.2))

states = nreps * [None]
result = parallel_execution(c, states)
```
will simulate the circuit for `nreps` different instances of the random numbers in the Pauli noise gates. The only issue with this is that currently the `parallel_execution` does not support measurements and getting the final states is probably not very useful for noise simulation.

I would try to implement this feature (noise simulation with parallel repeated execution) in a different PR after this is merged, I am just mentioning it here to make sure that passing a list of `None` as the initial state is okay in terms of the parallelization mechanism.